aws-lc-rs/src/agreement.rs

```diff
@@ -663,6 +661,136 @@ impl<B: AsRef<[u8]>> UnparsedPublicKey<B> {
+/// The format of a parsed public key.
+///
+/// This is used to distinguish between different types of public key formats
+/// supported by *aws-lc-rs*.
+#[non_exhaustive]
+pub enum ParsedPublicKeyFormat {
+    /// The key is in an X.509 SubjectPublicKeyInfo format.
+    X509,
+    /// The key is in an uncompressed form (X9.62).
+    Uncompressed,
+    /// The key is in a compressed form (SEC 1: Elliptic Curve Cryptography, Version 2.0).
+    Compressed,
+    /// The key is in a raw form. (X25519 only)
+    Raw,
+}
+
+/// A parsed public key for key agreement.
+impl ParsedPublicKey {
+    fn nid(&self) -> i32 {
+        self.nid
+    }
+
+    /// The format of the data the public key was parsed from.
+    #[must_use]
+    pub fn format(&self) -> ParsedPublicKeyFormat {
+        self.format
+    }
+
+    pub(crate) fn key(&self) -> &LcPtr<EVP_PKEY> {
+        &self.key
+    }
+
+    /// The algorithm of the public key.
+    #[must_use]
+    #[allow(non_upper_case_globals)]
+    pub fn alg(&self) -> &'static Algorithm {
+        match self.nid() {
+            NID_X25519 => &X25519,
+            NID_X9_62_prime256v1 => &ECDH_P256,
+            NID_secp384r1 => &ECDH_P384,
+            NID_secp521r1 => &ECDH_P521,
+            _ => unreachable!("Unreachable agreement algorithm nid: {}", self.nid()),
+        }
+    }
+}
+
+impl ParsedPublicKey {
+    #[allow(non_upper_case_globals)]
+    pub(crate) fn new(bytes: impl AsRef<[u8]>, nid: i32) -> Result<Self, KeyRejected> {
+        let key_bytes = bytes.as_ref();
+        if key_bytes.is_empty() {
+            return Err(KeyRejected::unspecified());
+        }
+        match nid {
+            NID_X25519 => {
+                let format: ParsedPublicKeyFormat;
+                let key = if let Ok(evp_pkey) =
+                    LcPtr::<EVP_PKEY>::parse_rfc5280_public_key(key_bytes, EVP_PKEY_X25519)
+                {
+                    format = ParsedPublicKeyFormat::X509;
+                    evp_pkey
+                } else {
+                    format = ParsedPublicKeyFormat::Raw;
+                    try_parse_x25519_public_key_raw_bytes(key_bytes)?
+                };
+
+                Ok(ParsedPublicKey { format, nid, key })
+            }
+            NID_X9_62_prime256v1 | NID_secp384r1 | NID_secp521r1 => {
+                let format: ParsedPublicKeyFormat;
+                let key = if let Ok(evp_pkey) =
+                    LcPtr::<EVP_PKEY>::parse_rfc5280_public_key(key_bytes, EVP_PKEY_EC)
+                {
+                    validate_ec_evp_key(&evp_pkey.as_const(), nid)?;
+                    format = ParsedPublicKeyFormat::X509;
+                    evp_pkey
+                } else if let Ok(evp_pkey) = parse_sec1_public_point(key_bytes, nid) {
+                    format = match key_bytes[0] {
+                        0x02 | 0x03 => ParsedPublicKeyFormat::Compressed,
+                        _ => ParsedPublicKeyFormat::Uncompressed,
```

Comment:
This matching is a bit inaccurate, `Uncompressed` formats typically only match to `0x04`. There's the possibility of `Hybrid` formats which are indicated with `0x06` or `0x07` and the default patch would match this to that.

I don't think we necessarily need to support `Hybrid` formats here if it's too much a hassle as they're rare and relatively unused, but we should definitely be a bit stricter here. Either adding an additional `ParsedPublicKeyFormat::Unknown` or directly failing here on a mismatch would also do?